jbpm-task-assigning/jbpm-task-assigning-core/src/test/java/org/jbpm/task/assigning/model/solver/realtime/BaseProblemFactChangeTest.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.task.assigning.model.solver.realtime;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+
+import org.jbpm.task.assigning.BaseTaskAssigningTest;
+import org.jbpm.task.assigning.model.TaskAssigningSolution;
+import org.junit.Rule;
+import org.junit.rules.ExpectedException;
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.impl.solver.ProblemFactChange;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public abstract class BaseProblemFactChangeTest extends BaseTaskAssigningTest {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    final Random random = new Random();
+
+    protected class ProgrammedProblemFactChange<C extends ProblemFactChange<TaskAssigningSolution>> {
+
+        private TaskAssigningSolution solutionAfterChange;
+
+        private C change;
+
+        public ProgrammedProblemFactChange() {
+        }
+
+        public ProgrammedProblemFactChange(C change) {
+            this.change = change;
+        }
+
+        public TaskAssigningSolution getSolutionAfterChange() {
+            return solutionAfterChange;
+        }
+
+        public void setSolutionAfterChange(TaskAssigningSolution solutionAfterChange) {
+            this.solutionAfterChange = solutionAfterChange;
+        }
+
+        public C getChange() {
+            return change;
+        }
+
+        public void setChange(C change) {
+            this.change = change;
+        }
+    }
+
+    protected TaskAssigningSolution executeSequentialChanges(TaskAssigningSolution solution, List<? extends ProgrammedProblemFactChange> changes) throws Exception {
+        Solver<TaskAssigningSolution> solver = createDaemonSolver();
+
+        //store the first solution that was produced by the solver for knowing how things looked like at the very
+        //beginning before any change was produced.
+        final TaskAssigningSolution[] initialSolution = {null};
+        final boolean[] changesInProgress = {false};
```

Comment:
Potential race condition.

Although this might be a viable solution for changing the flag inside the lambda, but the variable is accessed by two threads simultaneously. One way of dealing with this could be using `AtomicReference`:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReference.html

Reply:
Agreed, atomic reference is being used in other places where it's really needed. (will came up in next PR). But in this particular case by construction this potential concurrency scenario will never happen thanks to the use of programNextChange.acquire();

But no problem, it can be changed.